activation/activation.go

```
@@ -135,133 +139,69 @@ func (b *Builder) Start() {
+	<-b.syncer.Await()
+	challenge := &types.NIPSTChallenge{NodeId: b.nodeId}
+	if posAtx, err := b.GetPositioningAtx(); err != nil {
+		if !b.currentEpoch().IsGenesis() {
+			return fmt.Errorf("failed to get positioning ATX: %v", err)
 		}
-		prevAtxId = b.prevATX.Id()
-	}
-
-	posAtxEndTick := uint64(0)
-	b.posLayerID = types.LayerID(0)
-
-	//positioning atx is from this epoch as well, since we will be publishing the atx in the next epoch
-	//todo: what if no other atx was received in this epoch yet?
-	posAtxId := *types.EmptyAtxId
-	posAtx, err := b.GetPositioningAtx(epoch)
-	atxPubLayerID := types.LayerID(0)
-	if err == nil {
-		posAtxEndTick = posAtx.EndTick
-		b.posLayerID = posAtx.PubLayerIdx
-		atxPubLayerID = b.posLayerID.Add(b.layersPerEpoch)
-		posAtxId = posAtx.Id()
+		challenge.EndTick = b.tickProvider.NumOfTicks()
 	} else {
-		if !epoch.IsGenesis() {
-			return fmt.Errorf("cannot find pos atx in epoch %v: %v", epoch, err)
-		}
+		challenge.PositioningAtx = posAtx.Id()
+		challenge.PubLayerIdx = posAtx.PubLayerIdx.Add(b.layersPerEpoch)
+		challenge.StartTick = posAtx.EndTick
+		challenge.EndTick = posAtx.EndTick + b.tickProvider.NumOfTicks()
 	}
-
-	b.challenge = &types.NIPSTChallenge{
-		NodeId:               b.nodeId,
-		Sequence:             seq,
-		PrevATXId:            prevAtxId,
-		PubLayerIdx:          atxPubLayerID,
-		StartTick:            posAtxEndTick,
-		EndTick:              b.tickProvider.NumOfTicks(), //todo: add provider when
-		PositioningAtx:       posAtxId,
-		CommitmentMerkleRoot: commitmentMerkleRoot,
+	if prevAtx, err := b.GetPrevAtx(b.nodeId); err != nil {
+		challenge.CommitmentMerkleRoot = b.commitment.MerkleRoot
+	} else {
+		challenge.PrevATXId = prevAtx.Id()
+		challenge.Sequence = prevAtx.Sequence + 1
 	}
-
-	err = b.storeChallenge(b.challenge)
-	if err != nil {
-		log.Error("challenge cannot be stored: %s", err)
+	b.challenge = challenge
+	if err := b.storeChallenge(b.challenge); err != nil {
```

shouldn't you return this error?

---

I must admit I didn't think about this (this was the existing behaviour).

The risk if we keep the current behaviour is that the node registers to PoET with this challenge and then restarts at some point, essentially losing this epoch, since there's no ability to recover.

I guess it's better to fail fast. Good catch. Will change this.